## One Day Python Assignment

## Student Name: William Gardner

This is the assignment intended to teach Python from the basics all the way to using Pandas Dataframes. It is not intended to be comprehensive, but rather a survey of the major topics to get the learner from point A to B.    
   
Part 1 will be worth 100 points and Part 2 will be worth 100 points. There will only be one assignment on Canvas worth 100 points. The final grade for this will be divided by 2 and rounded up.     
So if you got 100 on Part 1 and 67 on Part 2, you will have 167 / 2 = 83.5. Extra credit is added onto Part 1, so, again: if you got 110 on Part 1 and 67 on Part 2, you will have 177 / 2 = 90.

### PART 1

 1. Print “Hello World”  (2 pts)

In [1]:
print("Hello World")

Hello World


2.  Save your age to a variable called “age”    
3.  Print “Hello World, I’m (age variable) today.”    (5 pts)

In [5]:
age = 27
print("Hello World, I'm", age, "today")

Hello World, I'm 27 today


4. Print “hello world” in all upper case using a Python string method to alter it (4 pts)

In [7]:
string = "hello world" #stores string in lowercase
s = string.upper() #converts string to uppercase
print(s) #prints the converted string

HELLO WORLD


 5. Save any decimal in a variable called a.   
 Cast a into an int and save it as b.  
 Cast a into a string and save it as c.   
 Print each variable and its type.   (11 pts)

In [9]:
a = 0.1
b = int(a) #casts a to int
c = str(a) #casts a to string
print(a, "is", type(a))
print(b, "is", type(b))
print(c, "is", type(c))

0.1 is <class 'float'>
0 is <class 'int'>
0.1 is <class 'str'>


6. Create a tuple of your favorite things to eat and print it.   (4 pts)

In [247]:
favorites = ("ice cream", "blueberries", "moros") #creates a tuple with favorite foods
print(favorites) #prints the tuple

('ice cream', 'blueberries', 'moros')


7. Create a dictionary named ‘classes’ with the name of the classes you are taking as the keys and the professors names as the values. 
Print the dictionary and tuple.   (10 pts)

In [248]:
classes = {
    "Principles of Programming Languages": "Professor Lloyd",
    "Introduction to Embedded Systems": "Professor Alhalabi",
    "Theory of Computation": "Professor Sorgente",
    "Engineering Design II": "Professor Zhong"
} #creates a dictionary of my course schedule with professors as values and classes as keys

print(classes)

{'Principles of Programming Languages': 'Professor Lloyd', 'Introduction to Embedded Systems': 'Professor Alhalabi', 'Theory of Computation': 'Professor Sorgente', 'Engineering Design II': 'Professor Zhong'}


8. Create a list called “whole” and put every number in it from 1 - 100  (10 pts)

In [262]:
whole = [] #creates an empty list
for i in range(1, 101): # loops from 1 to 100, stopping at 100, and appends the numbers to the list
    whole.append(i)
    
print(whole)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]


9. Create 4 empty lists called ‘div2’, ‘div3’, ‘div4’, and ‘div5’       
Create a loop that examines each number from 1 to 100 and:    
 If it is divisible by 2, put it in div2   (6 pts)    
 If it is divisible by 3, put it in div3   (6 pts)     
 If it is divisible by 4, put it in div4   (6 pts)    
 If it is divisible by 5, put it in div5   (6 pts)    
 Print all 5 of these lists

In [268]:
#creates 4 empty lists
div2 = []
div3 = []
div4 = []
div5 = []

for i in range (1, 101):
    if (i) % 2 == 0: #if divisible by 2
        div2.append(i) 
    if (i) % 3 == 0: #if divisible by 3
        div3.append(i)
    if (i) % 4 == 0: #if divisible by 4
        div4.append(i)
    if (i) % 5 == 0: #if divisible by 5
        div5.append(i)
        
print (div2, "\n", div3, "\n", div4, "\n", div5)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100] 
 [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99] 
 [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100] 
 [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100]


10. Create a new list called “divOver5”   (3 pts)

In [269]:
divOver5 = [] #creates a new list called divOver5

11. Create a new loop that goes through each number 1 through 100 and appends it to divOver5 if it is NOT IN div2 or div3 or div4 or div5      
 You must use logical operators here.   
Print divOver5   (10 points)    

In [270]:
for i in range(1, 101):
    if i not in div2 and i not in div3 and i not in div4 and i not in div5: #if not divisible by 2, 3, 4, or 5
        divOver5.append(i) #append to new list
        
print(divOver5)

[1, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 49, 53, 59, 61, 67, 71, 73, 77, 79, 83, 89, 91, 97]


12. Create a function called exp3 that takes an int x, raises it to the third power.     
Create a string in the function that says “x^3 = “ and concatenate x onto the end of the string.      
Now return the string from the function.  Call the function exp3 and print what it returns.   (15 pts)

In [272]:
def exp3(x):
    result = x**3 #stores x^3 in result
    strResult = "x^3 = " + str(result) #converts result to string and concatenates result to the end of the new string
    return strResult #returns the string result

exp3(5) #sample driver code

'x^3 = 125'

13. Iterate through the classes dictionary and print the keys    (5 pts)

In [110]:
for key in classes:
    print("Key is:", key) #iterates and prints the keys

Key is: Principles Programming Languages
Key is: Introduction to Embedded Systems
Key is: Theory of Computation
Key is: Engineering Design II


14. Iterate through the classes dictionary and print the values   (5 pts)

In [275]:
for key in classes:
    print("Value is:", classes[key]) #iterates and prints the values 

Value is: Professor Lloyd
Value is: Professor Alhalabi
Value is: Professor Sorgente
Value is: Professor Zhong


(Worth +10 points) Extra Points:   
Create a class called Student that has the properties:    
	name (holds student name)    
	age (holds student age)    
	birthmonth (hold students birthmonth)
	
	A count variable exists that is set to 0 (see lecture) and upon each call to the class this variable increments     

	2 functions in the class:  displayName which returns the name of the student and displayBirthmonth which returns the birth month of the student    

	Now create 3 students, use whatever data you want.       
	Call displayName() on student1     
	Call displayBirthmonth() on student2    
	print the student count    


In [276]:
class Student:
    count = 0 #creates a count variable and sets it to 0
    def __init__(self, name, age, birthmonth): #default constructor for the function, sets name to name, age to age, birthmonth to birthmonth
        self.name = name
        self.age = age
        self.birthmonth = birthmonth
        Student.count += 1 #each time a student is created, iterates count
    
    def displayName(self):
        print("The name of this student is", self.name) #prints the name of the student
        
    def displayBirthmonth(self):
        print("The birth month of this student", self.name, "is", self.birthmonth) #prints the birth month of the student
        
student1 = Student("Billy", 14, "January")
student2 = Student("Sally", 13, "February")
student3 = Student("Natalie", 12, "March")

student1.displayName() #calls displayName and prints name of Billy
student2.displayBirthmonth() #calls displayBirthmonth and prints birth month of Sally

print("There are currently", Student.count, "students in the class.") #displays count of students

    
    

The name of this student is Billy
The birth month of this student Sally is February
There are currently 3 students in the class.


### PART 2

This part will require you to use the film_results.txt file provided on Canvas. You will also need to import NumPy and Pandas.

1. Move the film_results.txt file from Blackboard into your project directory.     
 Create 3 lists, one for states, one for shows, and one for viewers.     
 Ingest data from text file and put it into a NumPy array (show_results.txt)      
	Itll look like a nested list.       
          [ [‘Oregon’, ‘Once Upon a Time’, ‘4075’] [...] …]     
 Print the raw data   (10 pts)    

In [289]:
import numpy as np, pandas as pd #imports necessary libraries

file_path = 'film_results.txt' #stores file path in easily-referenced variable

#creates three empty lists
states = []
shows = []
viewers = []

#opens the file in read-only mode, scans through it, and apends it line-by-line to a new list called data
with open(file_path, 'r') as file:
    data = []
    for line in file:
        lines = [element.strip() for element in line.split(',')]
        data.append(lines)

film_array = np.loadtxt(file_path, dtype=str, delimiter=',') #converts data to a numpy array
print("Raw Data:\n", data) #prints raw data
print("\nNumPy Array:\n", film_array) #prints numpy array

Raw Data:
 [['Oregon', 'Once Upon a Time', '4075'], ['Washington', 'Vikings', '3739'], ['Washington', 'Game of Thrones', '3997'], ['Nevada', 'Vikings', '1682'], ['Washington', 'Game of Thrones', '8983'], ['Washington', 'Vikings', '145'], ['Idaho', 'Daredevil', '2663'], ['California', 'Game of Thrones', '6361'], ['Idaho', 'Game of Thrones', '6788'], ['Idaho', 'The Flash', '3064'], ['Oregon', 'Gotham', '9389'], ['Washington', 'The Walking Dead', '4301'], ['Nevada', 'Game of Thrones', '7328'], ['California', 'The Flash', '9003'], ['Nevada', 'Gotham', '8575'], ['Idaho', 'Suits', '3239'], ['Idaho', 'Gotham', '6695'], ['Nevada', 'The Flash', '594'], ['Nevada', 'The Flash', '3793'], ['Nevada', 'Suits', '8544'], ['California', 'Vikings', '1059'], ['Nevada', 'Daredevil', '8200'], ['California', 'Daredevil', '1856'], ['Washington', 'Suits', '2871'], ['Washington', 'Better Call Saul', '9574'], ['Idaho', 'Vikings', '5218'], ['Idaho', 'The Walking Dead', '6256'], ['Nevada', 'Vikings', '8581'], ['Or

2.  Take the data from the NumPy array and sort it by state, show and viewers, putting each into the appropriate lists you defined earlier.  (so now you have 3 lists, one with states, one with shows and one with viewer counts.)  No duplicates in the states and shows.  Duplicates can and should exist in the viewers.      
So the states list will look like this: ['Washington', 'Nevada', 'Idaho', 'California', 'Oregon']      
 Print these unsorted lists    (5 pts)

In [290]:
# iterates through the numpy array, sorts it by state, show, and viewers, and appends to appropriate lists
# removes duplicates via not in methodology
for entry in film_array:
    state, show, viewer_count = entry
    if state not in states:
        states.append(state)
    if show not in shows:
        shows.append(show)
    viewers.append(viewer_count)

print("Unsorted lists:")
print("States:", states) #Note that states prints different than the example, as I have iterated from 0, where it is 0-indexed and Oregon is the first state to appear.
print("Shows:", shows)
print("Viewer counts:", viewers)


Unsorted lists:
States: ['Oregon', 'Washington', 'Nevada', 'Idaho', 'California']
Shows: ['Once Upon a Time', 'Vikings', 'Game of Thrones', 'Daredevil', 'The Flash', 'Gotham', 'The Walking Dead', 'Suits', 'Better Call Saul', 'Orange is the New Black']
Viewer counts: ['4075', '3739', '3997', '1682', '8983', '145', '2663', '6361', '6788', '3064', '9389', '4301', '7328', '9003', '8575', '3239', '6695', '594', '3793', '8544', '1059', '8200', '1856', '2871', '9574', '5218', '6256', '8581', '2460', '728', '7495', '2103', '8179', '1144', '7514', '271', '4353', '7647', '3408', '8443', '1970', '1711', '825', '7875', '1333', '9753', '4233', '1349', '1473', '4176', '2193', '8477', '6834', '2569', '8284', '4729', '4988', '1177', '9210', '6374', '9114', '9049', '6639', '5225', '4988', '2103', '6207', '6859', '1314', '6491', '5764', '960', '5205', '2241', '6930', '3699', '8447', '3784', '9469', '4746', '8430', '742', '9322', '3254', '5103', '7291', '8395', '7304', '2707', '2547', '5555', '6505', '18

3. Convert all 3 lists into NumPy arrays     
 print new NumPy Arrays    (10 pts)

In [293]:
#Converts all lists to numpy arrays
np_states = np.array(states)
np_shows = np.array(shows)
np_viewers = np.array(viewers)

#prints the new numpy arrays
print("States: ", np_states)
print("\nShows: ", np_shows)
print("\nViewers: ", np_viewers)


States:  ['Oregon' 'Washington' 'Nevada' 'Idaho' 'California']

Shows:  ['Once Upon a Time' 'Vikings' 'Game of Thrones' 'Daredevil' 'The Flash'
 'Gotham' 'The Walking Dead' 'Suits' 'Better Call Saul'
 'Orange is the New Black']

Viewers:  ['4075' '3739' '3997' '1682' '8983' '145' '2663' '6361' '6788' '3064'
 '9389' '4301' '7328' '9003' '8575' '3239' '6695' '594' '3793' '8544'
 '1059' '8200' '1856' '2871' '9574' '5218' '6256' '8581' '2460' '728'
 '7495' '2103' '8179' '1144' '7514' '271' '4353' '7647' '3408' '8443'
 '1970' '1711' '825' '7875' '1333' '9753' '4233' '1349' '1473' '4176'
 '2193' '8477' '6834' '2569' '8284' '4729' '4988' '1177' '9210' '6374'
 '9114' '9049' '6639' '5225' '4988' '2103' '6207' '6859' '1314' '6491'
 '5764' '960' '5205' '2241' '6930' '3699' '8447' '3784' '9469' '4746'
 '8430' '742' '9322' '3254' '5103' '7291' '8395' '7304' '2707' '2547'
 '5555' '6505' '1817' '7268' '8161' '517' '2439' '7332' '6149']


4. Sort the States and Shows arrays      
	Now your States array will look like:      
	['California', 'Idaho', 'Nevada', 'Oregon', 'Washington']      
Convert the Viewers array from STRINGS into INTS      
Sum up viewers list into one variable (you can do this in one line)       
Print: Sorted arrays (states and shows), viewers list (as ints), and the variable that is the sum of the viewers list.        (30 pts)

In [294]:
sorted_indices = np.lexsort((film_array[:, 2].astype(int), film_array[:, 1], film_array[:, 0])) #stores and sorts the list indicies in a numpy array
sorted_film_array = film_array[sorted_indices] #stores the sorted numpy array based on the indicies
unique_states = np.unique(sorted_film_array[:, 0]) #stores unique states in a numpy array
unique_shows = np.unique(sorted_film_array[:, 1]) #stores unique shows in a numpy array
viewers = sorted_film_array[:, 2].astype(int) #stores viewers in a numpy array
total_viewers = np.sum(viewers) #totals the viewers using np.sum method

#prints all arrays and data
print("States:", unique_states)
print("\nShows:", unique_shows)
print("\nViewers:", viewers)
print("\nTotal viewers:", total_viewers)

States: ['California' 'Idaho' 'Nevada' 'Oregon' 'Washington']

Shows: ['Better Call Saul' 'Daredevil' 'Game of Thrones' 'Gotham'
 'Once Upon a Time' 'Orange is the New Black' 'Suits' 'The Flash'
 'The Walking Dead' 'Vikings']

Viewers: [8430 1856 4746 7304  517 1177 6361 7332 1473 9753 7291 1314 2193 3408
 9049 3254 9003 4353 1059 8179 1817 2663  742 6788 4176 6695 6930 5225
 1333 3239 3064 6491 9114 6256 5218 7268  960 4233 5555 8200 5205 5764
 7328 7647 9469 8575 6639 1349 2707 8544  594 3793  825 1682 6859 7514
 8161 8581 9210 7875 9322 2439 6149 6374 2460 6505 8284 8477 9389 4075
 1144 2569 6207 2103 4729 5103 8395 8443 8447 1711 7495 1970 3784 9574
 2241 2547 3997 4988 8983 6834 2103 3699 2871  271 4988 4301  145  728
 3739]

Total viewers: 499923


5. Create 2 DataFrames:      
show_raw_stats: index = numpy sorted array of SHOWS; columns = numpy sorted array of STATES      
show_agg_stats: index = same as above; columns= a list with the words Max, Min, Totals and Percent in it (like this… [‘Max’,’Min’,’Totals’, ‘Percent’]        
Populate show_raw_stats with data from the Original Array ingested from film_results.txt. 
Populate the Max, Min, Totals, and Percent in show_agg_stats using the DataFrame native functionality    
Print both dataframe   (30 pts)     

HINT:  You will need to create a loop here that basically goes the length of the original array, and on each iteration, it grabs the STATE, SHOW, and VIEWERS number (itll be a string so youll have to convert it…).    
The final step of each iteration will be placing it in the dataframe in the correct spot as an accumulation. +=      (Otherwise you just writing over the value there)     


In [296]:

#creates an empty dataframe and populates it with shows and states, where shows are the index and columns are the states, with both index and columns being pre-sorted
show_raw_stats = pd.DataFrame(index=unique_shows, columns=unique_states, data=0)

#iterates through the numpy array, and adds the appropriate state, show, and viewers to the dataframe
#converts the viewer count to an integer
for entry in film_array:
    state, show, viewer_count = entry
    show_raw_stats.at[show, state] += int(viewer_count)

#creates and configures the show_agg_stats dataframe, with Max, Min, Totals, and Percent
#uses the show_raw_stats dataframe to calculate each via max, min, and sum methods
#calculates percent by aggregate data counts    
show_agg_stats = pd.DataFrame(index=unique_shows, columns=['Max', 'Min', 'Totals', 'Percent'])
show_agg_stats['Max'] = show_raw_stats.max(axis=1)
show_agg_stats['Min'] = show_raw_stats.min(axis=1)
show_agg_stats['Totals'] = show_raw_stats.sum(axis=1)
show_agg_stats['Percent'] = (show_agg_stats['Totals'] / show_agg_stats['Totals'].sum()) * 100

#prints headers and displays dataframes
print("show_raw_stats:")
display(show_raw_stats)
print("\nshow_agg_stats:")
display(show_agg_stats)

show_raw_stats:


California  Idaho  Nevada  Oregon  Washington
Better Call Saul               8430   8179    7268   17197       15328
Daredevil                     13906   4480   18948   14962        2241
Game of Thrones               15387   7530   35413    2460       20515
Gotham                         1473  10871    8575   32655        6834
Once Upon a Time               9753   6930    6639    4075        2103
Orange is the New Black        7291   5225    1349    1144        3699
Suits                         15964   4572   11251    8776        2871
The Flash                     12257  18669    4387   28773        5259
The Walking Dead               4353   6256     825    8447        4301
Vikings                        1059   5218   42007    9206        4612


show_agg_stats:


Max   Min  Totals    Percent
Better Call Saul         17197  7268   56402  11.282137
Daredevil                18948  2241   54537  10.909080
Game of Thrones          35413  2460   81305  16.263505
Gotham                   32655  1473   60408  12.083461
Once Upon a Time          9753  2103   29500   5.900909
Orange is the New Black   7291  1144   18708   3.742176
Suits                    15964  2871   43434   8.688138
The Flash                28773  4387   69345  13.871136
The Walking Dead          8447   825   24182   4.837145
Vikings                  42007  1059   62102  12.422313

6. Print the answer to these questions in a Markdown cell, do not print them with Python:        
Which Show has the highest percentage?   (5 pts)      
Which Show has the lowest percentage?    (5 pts)     
Which show is your favorite?             (5 pts)     

The show with the highest percentage is Game of Thrones, at roughly 16.26%.

The show with the lowest percentage is Orange is the New Black, at roughly 3.74%.

My favorite of these shows is Game of Thrones (although the ending was not great...).

### Please try to follow the directions as closely as possible.  You might think of better ways to implement this, with less lines of code the skips steps and thats great, but not for this assignment.  If you do think of these, put them in comments and extra points will be given for those but follow these steps.